### PR TITLE
[SPARK-46112][BUILD][PYTHON] Implement lint check for PySpark custom errors

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to introduce a new script, `check_pyspark_custom_errors.py`, integrated into the `dev/lint-python` process. This script is designed to enforce the use of PySpark's custom error classes.

### Why are the changes needed?

To keep the consistency of errors generated from PySpark systematically.

### Does this PR introduce _any_ user-facing change?

No. It is a development tool enhancement to ensure code quality and consistency in error handling within the PySpark codebase.

For example, the `lint-python` script will detect the exact location where PySpark custom error is not used and generate the following error:

```
(pyspark-dev-env) haejoon.lee@NQ679Q495F spark % ./dev/lint-python
starting python compilation test...
python compilation succeeded.

starting black test...
black checks passed.

starting PySpark custom errors check...
PySpark custom errors check failed!

PySpark custom errors check found issues in the following files:
python/pyspark/sql/connect/utils.py:48: raise ImportError(
python/pyspark/sql/connect/utils.py:52: raise ImportError(
python/pyspark/sql/connect/utils.py:65: raise ImportError(
python/pyspark/sql/connect/utils.py:77: raise ImportError(
python/pyspark/sql/connect/client/retries.py:236: raise RetriesExceeded from exception
python/pyspark/sql/pandas/serializers.py:741: raise ValueError(
python/pyspark/sql/pandas/utils.py:35: raise ImportError(
python/pyspark/sql/pandas/utils.py:39: raise ImportError(
python/pyspark/sql/pandas/utils.py:60: raise ImportError(
python/pyspark/sql/pandas/utils.py:65: raise ImportError(

Please use PySpark custom errors defined in pyspark.errors.
```

### How was this patch tested?

Run `dev/lint-python` and see "PySpark custom errors check passed."
```
(pyspark-dev-env) haejoon.lee@NQ679Q495F spark % ./dev/lint-python
starting python compilation test...
python compilation succeeded.

starting black test...
black checks passed.

starting PySpark custom errors check...
PySpark custom errors check passed.
```

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
I wonder if there are other thridparty linter integration for this ..

**Comment by itholic** [comment]
At least from my research before writing the script by my own, it seems there are no existing third-party linters that specifically address our custom error checking requirements in PySpark. IMHO the current implementation appears to be the most viable solution for now. But please let me know if there is any alternate way to support it!

File: dev/check_pyspark_custom_errors.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    TARGET_PATHS = ["python/pyspark/sql"]
+    EXCLUDE_PATHS = [
+        "python/pyspark/sql/tests",
+        "python/pyspark/sql/connect/proto",
+    ]
+
+    # Check errors
+    errors_found = check_pyspark_custom_errors(TARGET_PATHS, EXCLUDE_PATHS)
+    if errors_found:
+        print("\nPySpark custom errors check found issues in the following files:")
```
Comment (HyukjinKwon) [inline dev/check_pyspark_custom_errors.py:96]: print it out to stderr

File: dev/check_pyspark_custom_errors.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        Path to the file to check.
+    pyspark_error_list : list of str
+        List of PySpark-specific error names.
+    """
+    errors_found = []
+    with open(file_path, "r") as file:
+        for line_num, line in enumerate(file, start=1):
+            if line.strip().startswith("raise"):
+                parts = line.split()
+                if len(parts) > 1 and parts[1][0].isupper():
```
Comment (HyukjinKwon) [inline dev/check_pyspark_custom_errors.py:51]: Can you add a comment which case this searches for?

File: dev/check_pyspark_custom_errors.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+def check_pyspark_custom_errors(target_paths, exclude_paths):
+    """
+    Check PySpark-specific errors in multiple paths.
+
+    Parameters
+    ----------
+    target_paths : list of str
+        List of paths to check for PySpark-specific errors.
+    exclude_paths : list of str
+        List of paths to exclude from the check.
```
Comment (HyukjinKwon) [inline dev/check_pyspark_custom_errors.py:66]: describe `Returns`

File: dev/check_pyspark_custom_errors.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+def check_errors_in_file(file_path, pyspark_error_list):
+    """
+    Check if a file uses PySpark-specific errors correctly.
+
+    Parameters
+    ----------
+    file_path : str
+        Path to the file to check.
+    pyspark_error_list : list of str
+        List of PySpark-specific error names.
```
Comment (HyukjinKwon) [inline dev/check_pyspark_custom_errors.py:44]: Returns

File: dev/check_pyspark_custom_errors.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
+import sys
```
Comment (HyukjinKwon) [inline dev/check_pyspark_custom_errors.py:1]: add license

File: dev/lint-python [hunk 6]
```diff
@@ -46,6 +46,9 @@ while (( "$#" )); do
     --black)
       BLACK_TEST=true
       ;;
+    --custom-pyspark-error)
```
[the remark this answers is not in the document]
Comment (itholic) [reply dev/lint-python:49]: Fixed

File: dev/check_pyspark_custom_errors.py (hunk 2 above)
Comment (itholic) [reply dev/check_pyspark_custom_errors.py:79]: Sure. Added `# Check for 'raise' statement and ensure the error raised is a capitalized word.`

File: dev/check_pyspark_custom_errors.py (hunk 5 above)
Comment (itholic) [reply dev/check_pyspark_custom_errors.py:20]: Added license.

File: dev/check_pyspark_custom_errors.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        "python/pyspark/sql/connect/proto",
+    ]
+
+    # Check errors
+    errors_found = check_pyspark_custom_errors(TARGET_PATHS, EXCLUDE_PATHS)
+    if errors_found:
+        print("\nPySpark custom errors check found issues in the following files:")
+        for error in errors_found:
+            print(error)
+        print("\nPlease use PySpark custom errors defined in pyspark.errors.")
```
[the remark this answers is not in the document]
Comment (itholic) [reply dev/check_pyspark_custom_errors.py]: Rephrase:

```diff
-        print("\nPlease use PySpark custom errors defined in pyspark.errors.")
+        print("\nUse existing or create a new custom error from pyspark.errors.", file=sys.stderr)
```

**Review by HyukjinKwon — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
Nice!

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 4.

**Pull request closed without merging**